compiler/codegen/symbol.cpp

```diff
@@ -1110,18 +1112,45 @@ std::string ArgSymbol::getPythonArgTranslation() {
   std::string strname = cname;
 
   if (t == dtStringC) {
-    std::string res = "\tcdef const char* chpl_" + strname + " = " + strname;
-    res += "\n";
+    std::string res;
+
+    // First, check to see if the input type is the Python "str" type.
+    res += "\tif type(" + strname + ") != str:\n";
+    res += "\t\traise TypeError(\"Expected \'str\' in conversion to ";
+    res += " \'const char*\', ";
+    res += "found \" + str(type(" + strname + ")))\n";
+
+    // Encode string as bytes. We still hold a ref.
+    res += "\t" + strname + " = " + strname + ".encode(\'utf-8\')\n";
+
+    // Cython will auto convert the bytes to "char*";
+    res += "\tcdef char* chpl_" + strname + " = " + strname + "\n";
+
     return res;
   } else if (t->getValType() == exportTypeChplBytesWrapper) {
+    Type* origt = getUnwrappedArg(this)->type->getValType();
+    INT_ASSERT(origt == dtBytes || origt == dtString);
+  
+    const char* pythonType = (origt == dtBytes) ? "bytes" : "str";
     std::string res;
 
-    // First, check to see if the input type is the Python "bytes" type.
-    res += "\tif type(" + strname + ") != bytes:\n";
-    res += "\t\traise TypeError(\"Expected \'bytes\' in conversion to ";
-    res += " \'chpl_bytes_wrapper\', ";
+    //
+    // Generate a Python TypeError if the Python type does not match the
+    // Chapel type (string or bytes).
+    //
+    res += "\tif type(" + strname + ") != ";
+    res += pythonType;
+    res += ":\n";
+    res += "\t\traise TypeError(\"Expected \'";
+    res += pythonType;
+    res += "\' in conversion to \'chpl_byte_buffer\', ";
```

Comment:
I probably should have caught this when it was originally added, but it might be nice to give an error using the Chapel type name instead of our under-the-covers conversion type name

Reply:
That's a good idea.